File: packages/sdk-py/af_stack/cost.py
"""``suite.cost.*`` — cost event log browsing.

Maps to the cost event REST surface in
``apps/dashboard/src/lib/api.ts``:

==================================  =========================================
SDK call                            Endpoint
==================================  =========================================
:func:`events`                      ``GET    /api/v1/cost/events``
==================================  =========================================

Each :class:`CostEvent` is one billable LLM call written by the cost
recorder (Phase 7.2). The dashboard renders these rows in the live cost
log, and a :class:`CostEventList` envelopes the paginated response.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from . import _http


class CostEvent(BaseModel):
    """One billable LLM call.

    Mirrors ``CostEventSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    id: str
    tenant_id: str | None = None
    api_key_id: str | None = None
    model: str
    provider: str
    agent: str | None = None
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    cost_usd: float
    cached: bool
    latency_ms: int
    occurred_at: str


class CostEventList(BaseModel):
    """Paginated list of cost events.

    Mirrors ``CostEventListSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    events: list[CostEvent] = Field(default_factory=list)
    total: int = 0
    has_more: bool = False


async def events(
    *,
    tenant: str | None = None,
    model: str | None = None,
    from_: str | datetime | None = None,
    to: str | datetime | None = None,
    limit: int = 100,
    offset: int = 0,
) -> CostEventList:
    """List cost events with optional filters.

    ``from_`` and ``to`` accept RFC3339 strings or :class:`datetime`
    instances. The trailing-underscore name avoids a clash with the
    ``from`` keyword; the wire param is plain ``from``.
    """
    if limit <= 0:
        raise ValueError("limit must be a positive integer")
    if offset < 0:
        raise ValueError("offset must be non-negative")
    params: dict[str, Any] = {"limit": limit, "offset": offset}
    if tenant:
        params["tenant"] = tenant
    if model:
        params["model"] = model
    if from_ is not None:
        params["from"] = _to_iso(from_)
    if to is not None:
        params["to"] = _to_iso(to)
    body = await _http.request_json("GET", "/cost/events", params=params)
    return CostEventList.model_validate(
        body or {"events": [], "total": 0, "has_more": False}
    )


def _to_iso(value: str | datetime) -> str:
    """Serialise a :class:`datetime` to RFC3339; pass strings through.

    Naive datetimes are treated as UTC for predictability — same
    convention used by :mod:`af_stack.jobs` and :mod:`af_stack.admin.audit`.
    """
    if isinstance(value, datetime):
        iso = value.isoformat()
        if value.tzinfo is None:
            return iso + "Z"
        return iso
    return value


__all__ = ["CostEvent", "CostEventList", "events"]

File: packages/sdk-ts/src/cost.ts
// suite.cost.* — cost event log browsing.
//
// Endpoint (mirroring `apps/dashboard/src/lib/api.ts`):
//
//   GET /api/v1/cost/events?tenant=&model=&from=&to=&limit=&offset=
//
// Each `CostEvent` is one billable LLM call written by the cost recorder
// (Phase 7.2). Schemas exposed here are camelCase; the wire stays
// snake_case and is translated via `toCamel`.

import { z } from "zod"
import { request, toCamel, type HttpOptions } from "./_http.js"

// ---------- shared schemas (mirror lib/api.ts, camelCase) ----------

export const CostEventSchema = z.object({
  id: z.string(),
  tenantId: z.string().nullable(),
  apiKeyId: z.string().nullable(),
  model: z.string(),
  provider: z.string(),
  agent: z.string().nullable(),
  promptTokens: z.number(),
  completionTokens: z.number(),
  totalTokens: z.number(),
  costUsd: z.number(),
  cached: z.boolean(),
  latencyMs: z.number(),
  occurredAt: z.string(),
})
export type CostEvent = z.infer<typeof CostEventSchema>

export const CostEventListSchema = z.object({
  events: z.array(CostEventSchema),
  total: z.number(),
  hasMore: z.boolean(),
})
export type CostEventList = z.infer<typeof CostEventListSchema>

// ---------- public API ----------

export interface ListCostEventsOptions extends HttpOptions {
  tenant?: string
  model?: string
  /** RFC3339 timestamp or Date. */
  from?: string | Date
  /** RFC3339 timestamp or Date. */
  to?: string | Date
  limit?: number
  offset?: number
}

/** List cost events with optional filters. */
export async function events(
  opts: ListCostEventsOptions = {},
): Promise<CostEventList> {
  const {
    tenant,
    model,
    from,
    to,
    limit = 100,
    offset = 0,
    ...http
  } = opts
  if (limit <= 0) throw new Error("limit must be a positive integer")
  if (offset < 0) throw new Error("offset must be non-negative")
  const qs = new URLSearchParams()
  if (tenant !== undefined && tenant !== "") qs.set("tenant", tenant)
  if (model !== undefined && model !== "") qs.set("model", model)
  if (from !== undefined) qs.set("from", toIso(from))
  if (to !== undefined) qs.set("to", toIso(to))
  qs.set("limit", String(limit))
  qs.set("offset", String(offset))
  const raw = await request<unknown>(
    "GET",
    `/cost/events?${qs.toString()}`,
    null,
    http,
  )
  return CostEventListSchema.parse(toCamel(raw))
}

function toIso(value: string | Date): string {
  return value instanceof Date ? value.toISOString() : value
}

/** Namespace object — the shape `suite.cost` is built from. */
export const cost = {
  events,
} as const
